src/providers/postgres/qgspostgresprovider.cpp

```diff
@@ -345,6 +354,45 @@ void QgsPostgresProvider::setTransaction( QgsTransaction *transaction )
   mTransaction = static_cast<QgsPostgresTransaction *>( transaction );
 }
 
+QgsPostgresProvider::PostgresReferencedGeometryConverter::PostgresReferencedGeometryConverter( QMap<int, QgsCoordinateReferenceSystem> *crsCache,
+    QMutex *crsCacheMutex )
+  : mCrsCache( crsCache )
+  , mCrsCacheMutex( crsCacheMutex )
+{
+}
+
+QgsPostgresProvider::PostgresReferencedGeometryConverter *QgsPostgresProvider::PostgresReferencedGeometryConverter::clone() const
+{
+  return new PostgresReferencedGeometryConverter( *this );
+}
+
+QgsReferencedGeometry QgsPostgresProvider::PostgresReferencedGeometryConverter::stringToReferencedGeometry( const QString &referencedGeometryString ) const
+{
+  thread_local const QRegularExpression regularExpressionSRID( "^SRID=(\\d+);" );
+
+  QRegularExpressionMatch regularExpressionMatch = regularExpressionSRID.match( referencedGeometryString );
+  if ( !regularExpressionMatch.hasMatch() )
+    return QgsReferencedGeometry();
+
+  QString wkt = referencedGeometryString.mid( regularExpressionMatch.captured( 0 ).size() );
+  int srid = regularExpressionMatch.captured( 1 ).toInt();
+
+  QgsGeometry geom = QgsGeometry::fromWkt( wkt );
+
+  // Look if crs already cached
+  QMutexLocker locker( mCrsCacheMutex );
+  if ( mCrsCache->contains( srid ) )
```

Comment:
@nyalldawson I passed the crs cache to the converter object. In most cases the right crs will already be in the cache at the time of rendering labels.
Do you think this could be a good enough solution? Otherwise i will embrace your suggestion about splitting the pr

Reply:
No, that doesn't help -- it uses a different definition of IDs. (And there's a chance a user will have two postgis connections with different crs definitions for the same I'd number)

Reply:
Ok i understand, but in this case the current implementation is also wrong. The crs cache is static and not per provider